engine/src/types.rs

```diff
@@ -299,6 +299,17 @@ impl<'a> Array<'a> {
         arr
     }
 
+    pub(crate) fn as_ref(&'a self) -> Array<'a> {
```

Comment:
These are pretty bad - unlike most `as_ref`, they're not merely re-wrapping references (cheap) but actually recreate structures (expensive).

Reply:
I agree, we should change the name. In the mean time, I tried to change `Array` / `Map` type to `Cow<'a, Array>` / `Cow<'a, Map>` without much luck for now.